server/index.js

```diff
@@ -447,30 +435,12 @@ io.on('connection', function(socket) {
             return;
         }
 
-        runAndCatchErrors(game, () => {           
-            game.playerLeave(socket.request.user.username, 'has left the game');
-
-            var player = game.players[socket.request.user.username];
-
-            if(!player) {
-                return;
-            }
-
-            if(game.isSpectator(player)) {
-                delete game.players[socket.request.user.username];
-                io.to(player.id).emit('gamestate', game.getState(player.name));
-            } else {
-                game.players[socket.request.user.username].left = true;
-
-                if(game.started && !game.finishedAt) {
-                    game.finishedAt = new Date();
-                    game.saveGame();
-                }                
-            }
+        runAndCatchErrors(game, () => {
+            game.leave(socket.request.user.username);
+            // Force update of game state to make sure spectators get disconnected.
+            socket.emit('gamestate', game.getState(socket.request.user.username));
```

Comment:
Most of this refactor was just pushing existing code around, but this is an actual change. This used to only happen if a spectator left, but I didn't want to send in the socket / io object into game to only do it for spectators. I think it's fine (brief testing didn't come up with anything) but please review carefully.